velox/connectors/hive/storage_adapters/s3fs/tests/S3FileTest.cpp

```diff
@@ -0,0 +1,151 @@
+    writeFile->append("ddddd");
+    ASSERT_EQ(writeFile->size(), 15 + kOneMB);
+  }
+
+  void readData(ReadFile* readFile) {
+    Arena arena;
+    ASSERT_EQ(readFile->size(), 15 + kOneMB);
+    ASSERT_EQ(readFile->pread(10 + kOneMB, 5, &arena), "ddddd");
+    ASSERT_EQ(readFile->pread(0, 10, &arena), "aaaaabbbbb");
+    ASSERT_EQ(readFile->pread(10, kOneMB, &arena), std::string(kOneMB, 'c'));
+    ASSERT_EQ(readFile->size(), 15 + kOneMB);
+    const std::string_view arf = readFile->pread(5, 10, &arena);
+    const std::string zarf = readFile->pread(kOneMB, 15);
+    auto buf = std::make_unique<char[]>(8);
+    const std::string_view warf = readFile->pread(4, 8, buf.get());
+    const std::string_view warfFromBuf(buf.get(), 8);
+    ASSERT_EQ(arf, "bbbbbccccc");
+    ASSERT_EQ(zarf, "ccccccccccddddd");
+    ASSERT_EQ(warf, "abbbbbcc");
+    ASSERT_EQ(warfFromBuf, "abbbbbcc");
+  }
+
+  static std::shared_ptr<MinioServer> minioServer_;
+};
+
+std::shared_ptr<MinioServer> S3FileSystemTest::minioServer_ = nullptr;
+
+TEST_F(S3FileSystemTest, writeAndRead) {
+  const char* bucket = "data";
+  const char* file = "test.txt";
+  const std::string filename = getLocalPath(bucket) + "/" + file;
+  const std::string s3File = getBucket(bucket) + "/" + file;
+  addBucket(bucket);
+  {
+    LocalWriteFile writeFile(filename);
+    writeData(&writeFile);
+  }
+  auto hiveConnectorConfigs = minioServer_->getHiveConfig();
+  std::shared_ptr<const Config> config =
+      std::make_shared<const core::MemConfig>(std::move(hiveConnectorConfigs));
+  filesystems::S3FileSystem s3fs(config);
+  s3fs.initializeClient();
+  auto readFile = s3fs.openFileForRead(s3File);
+  readData(readFile.get());
+}
+
+TEST_F(S3FileSystemTest, missingFile) {
+  const char* bucket = "data1";
+  const char* file = "i-do-not-exist.txt";
+  const std::string s3File = getBucket(bucket) + "/" + file;
+  addBucket(bucket);
+  auto hiveConnectorConfigs = minioServer_->getHiveConfig();
+  std::shared_ptr<const Config> config =
+      std::make_shared<const core::MemConfig>(std::move(hiveConnectorConfigs));
+  filesystems::S3FileSystem s3fs(config);
+  s3fs.initializeClient();
+  EXPECT_THROW(s3fs.openFileForRead(s3File), VeloxException);
+}
+
+TEST_F(S3FileSystemTest, viaRegistry) {
+  const char* bucket = "data2";
+  const char* file = "test.txt";
+  const std::string filename = getLocalPath(bucket) + "/" + file;
+  const std::string s3File = getBucket(bucket) + "/" + file;
+  filesystems::registerS3FileSystem();
+  addBucket(bucket);
+  {
+    LocalWriteFile writeFile(filename);
+    writeData(&writeFile);
+  }
+  auto hiveConnectorConfigs = minioServer_->getHiveConfig();
+  std::shared_ptr<const Config> config =
+      std::make_shared<const core::MemConfig>(std::move(hiveConnectorConfigs));
+  auto s3fs = filesystems::getFileSystem(s3File, config);
+  auto readFile = s3fs->openFileForRead(s3File);
+  readData(readFile.get());
+}
+
+TEST(S3UtilTest, utilTest) {
```

Comment:
Would you move it into a separate test file?

nit: utilTest -> basic or break it down into multiple tests, e.g. isS3File, getS3Path, getBucketAndKeyFromS3Path